app/rdf/query-hierarchies.ts

```diff
@@ -73,33 +73,29 @@ const toTree = (
   return sortChildren(results.map((r) => serializeNode(r, 0)).filter(truthy));
 };
 
-const findHierarchiesForDimension = (
-  cube: Cube,
+const getDimensionHierarchies = async (
+  cube: ExtendedCube,
   dimensionIri: string,
   locale: string
 ) => {
-  const newHierarchies = uniqBy(
-    cube.ptr
-      .any()
-      .has(ns.sh.path, rdf.namedNode(dimensionIri))
-      .has(ns.cubeMeta.inHierarchy)
-      .out(ns.cubeMeta.inHierarchy)
-      .toArray(),
-    (x) => getName(x, locale)
-  );
-  if (newHierarchies) {
-    return newHierarchies;
+  const cubeIri = cube.term?.value;
+
+  if (!cubeIri) {
+    throw new Error("Cube must have an iri!");
   }
-  const legacyHierarchies = cube.ptr
+
+  // Can't rely on the cube here, because sometimes it might contain duplicate
+  // hierarchies, due to fetching both cube and shape when cube is initialized.
+  // Here we are only interested in shape hierarchies, that's why we don't have
+  // to initialize the cube, but rather just fetch its shape.
+  const hierarchies = cube.shapePtr
     .any()
     .has(ns.sh.path, rdf.namedNode(dimensionIri))
-    .has(ns.cubeMeta.hasHierarchy)
-    .out(ns.cubeMeta.hasHierarchy)
+    .has(ns.cubeMeta.inHierarchy)
+    .out(ns.cubeMeta.inHierarchy)
     .toArray();
-  if (legacyHierarchies) {
-    return legacyHierarchies;
-  }
-  return [];
+
+  return uniqBy(hierarchies, (hierarchy) => getName(hierarchy, locale));
```

Comment:
For the splitted issue : getDimensionHierarchies could be on ExtendedCube. This way we will hide the complexity of the shapePtr there.